data/scripts.js

```diff
@@ -292,11 +255,290 @@ let BattleScripts = {
 			return false;
 		}
 
+		return true;
+	},
+	trySpreadMoveHit(targets, pokemon, move) {
+		if (!targets.length) {
+			this.attrLastMove('[notarget]');
+			this.add(this.gen >= 5 ? '-fail' : '-notarget', pokemon);
+			return false;
+		}
+		if (targets.length > 1) move.spreadHit = true;
+
+		let moveSteps = [
+			// 0. check for semi invulnerability
+			this.tryImmunityEvent,
+
+			// 1. check for type immunity (this is step 2 in gen 7+)
+			this.typeImmunity,
+
+			// 2. run the 'TryHit' event (Protect, Magic Bounce, Volt Absorb, etc.) (this is step 1 in gen 7)
+			this.tryHitEvent,
+
+			// 3. check for powder immunity
+			this.powderImmunity,
+
+			// 4. check for prankster immunity
+			this.pranksterImmunity,
+
+			// 5. check accuracy
+			this.accuracy,
+
+			// 6. break protection effects
+			this.breakProtect,
+
+			// 7. steal positive boosts (Spectral Thief)
+			this.stealBoosts,
+
+			// 8. loop that processes each hit of the move (has its own steps per iteration)
+			this.moveHitLoop,
+
+			// 9. effects that run after secondary effects, such as Color Change
+			this.afterMoveSecondaryEvent,
+		];
+		if (this.gen >= 7) {
+			// Swap step 1 with step 2
+			moveSteps.splice(2, 0, moveSteps.splice(1, 1)[0]);
+		}
+
+		this.setActiveMove(move, pokemon, targets[0]);
+		move.zBrokeProtect = false;
+
+		let hitResult = this.singleEvent('PrepareHit', move, {}, targets[0], pokemon, move);
+		if (!hitResult) {
+			if (hitResult === false) {
+				this.add('-fail', pokemon);
+				this.attrLastMove('[still]');
+			}
+			return false;
+		}
+		this.runEvent('PrepareHit', pokemon, targets[0], move);
+
+		if (!this.singleEvent('Try', move, null, pokemon, targets[0], move)) {
+			return false;
+		}
+
+		let finalResult;
+		let hitResults;
+		for (const step of moveSteps) {
+			hitResults = step(targets, pokemon, move);
+			if (!hitResults) continue;
+			for (let i = 0; i < targets.length; i++) {
+				if (!hitResults[i]) {
+					// hit resolved for this target.
+					// remove target and corresponding hitResults entry from their respective arrays
+					targets.splice(i, 1);
+					let targetResult = hitResults.splice(i, 1)[0];
+					i--;
+					// store failure/NOT_FAILURE in finalResult, with the following priority:
+					// truthy/number > false > '' > undefined
+					if (targetResult || targetResult === 0) {
+						finalResult = targetResult;
+					} else if (finalResult !== false || targetResult !== this.NOT_FAILURE) {
+						finalResult = targetResult;
+					}
+				}
+			}
+		}
+
+		let moveResult = !!targets.length;
+		if (!moveResult && finalResult === this.NOT_FAILURE) pokemon.moveThisTurnResult = null;
+		if (move.spreadHit) this.attrLastMove('[spread] ' + targets.join(','));
+		return moveResult;
+	},
+	tryImmunityEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryImmunity', targets, pokemon, move);
```

Comment:
It crashes here, claiming that `this` is undefined. I have no idea how that’s possible, since you can’t even call this function without a defined `this`.

Reply:
...oh, no, you can. I just realized it now.

Reply:
The problem is the way I called the functions from the array.

Reply:
What does crash mean?

An actual uncaught exception? Do you have the stack trace? The stack trace should make it obvious what's going on.